### DISCUSSION TITLE
How to get user for audit?

### DISCUSSION THREAD

**Discussion by newave-ws** [owen-it/laravel-auditing#945] [Q&A] [answered]
I trying to get the relative user of the audit?

this is the audit.php
`<?php

return [

    'enabled' => env('AUDITING_ENABLED', false),

    /*
    |--------------------------------------------------------------------------
    | Audit Implementation
    |--------------------------------------------------------------------------
    |
    | Define which Audit model implementation should be used.
    |
    */

    'implementation' => OwenIt\Auditing\Models\Audit::class,

    /*
    |--------------------------------------------------------------------------
    | User Morph prefix & Guards
    |--------------------------------------------------------------------------
    |
    | Define the morph prefix and authentication guards for the User resolver.
    |
    */

    'user'      => [
        'morph_prefix' => 'user',
        'guards'       => [
            'web',
            'api'
        ],
        'resolver'     => OwenIt\Auditing\Resolvers\UserResolver::class
    ],

    /*
    |--------------------------------------------------------------------------
    | Audit Resolvers
    |--------------------------------------------------------------------------
    |
    | Define the IP Address, User Agent and URL resolver implementations.
    |
    */
    'resolvers' => [
        'ip_address' => OwenIt\Auditing\Resolvers\IpAddressResolver::class,
        'user_agent' => OwenIt\Auditing\Resolvers\UserAgentResolver::class,
        'url'        => OwenIt\Auditing\Resolvers\UrlResolver::class,
    ],

    /*
    |--------------------------------------------------------------------------
    | Audit Events
    |--------------------------------------------------------------------------
    |
    | The Eloquent events that trigger an Audit.
    |
    */

    'events' => [
        'created',
        'updated',
        'deleted',
        'restored'
    ],

    /*
    |--------------------------------------------------------------------------
    | Strict Mode
    |--------------------------------------------------------------------------
    |
    | Enable the strict mode when auditing?
    |
    */

    'strict' => false,

    /*
    |--------------------------------------------------------------------------
    | Global exclude
    |--------------------------------------------------------------------------
    |
    | Have something you always want to exclude by default? - add it here.
    | Note that this is overwritten (not merged) with local exclude
    |
    */

    'exclude' => [],

    /*
    |--------------------------------------------------------------------------
    | Empty Values
    |--------------------------------------------------------------------------
    |
    | Should Audit records be stored when the recorded old_values & new_values
    | are both empty?
    |
    | Some events may be empty on purpose. Use allowed_empty_values to exclude
    | those from the empty values check. For example when auditing
    | model retrieved events which will never have new and old values.
    |
    |
    */

    'empty_values'         => true,
    'allowed_empty_values' => [
        'retrieved'
    ],

    /*
    |--------------------------------------------------------------------------
    | Audit Timestamps
    |--------------------------------------------------------------------------
    |
    | Should the created_at, updated_at and deleted_at timestamps be audited?
    |
    */

    'timestamps' => false,

    /*
    |--------------------------------------------------------------------------
    | Audit Threshold
    |--------------------------------------------------------------------------
    |
    | Specify a threshold for the amount of Audit records a model can have.
    | Zero means no limit.
    |
    */

    'threshold' => 0,

    /*
    |--------------------------------------------------------------------------
    | Audit Driver
    |--------------------------------------------------------------------------
    |
    | The default audit driver used to keep track of changes.
    |
    */

    'driver' => 'database',

    /*
    |--------------------------------------------------------------------------
    | Audit Driver Configurations
    |--------------------------------------------------------------------------
    |
    | Available audit drivers and respective configurations.
    |
    */

    'drivers' => [
        'database' => [
            'table'      => 'audits',
            'connection' => null,
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Audit Queue Configurations
    |--------------------------------------------------------------------------
    |
    | Available audit queue configurations.
    |
    */

    'queue' => [
        'enable'     => false,
        'connection' => 'sync',
        'queue'      => 'default',
        'delay'      => 0,
    ],

    /*
    |--------------------------------------------------------------------------
    | Audit Console
    |--------------------------------------------------------------------------
    |
    | Whether console events should be audited (eg. php artisan db:seed).
    |
    */

    'console' => false,
];
`

And this is the function that i'm trying to get the audits

`$modelClass = 'App\\Models\\' . $request->model;
        $modelInstance = resolve($modelClass);
        $query = $modelInstance->query()->where('id', $request->id)
            ->with('audits', function ($query) {
                $query->orderBy('id', 'desc')
                    ->select('id', 'auditable_id', 'auditable_type', 'user_id', 'event', 'new_values', 'old_values', 'ip_address', 'user_agent', 'created_at')
                    ->with('user');
            });`

the response is the getting null user field

**Comment by willpower232** [comment]
What do the rows in your audit table look like? Are auditable_type and auditable_id filled in?

**Reply by newave-ws** [reply]
Yes, i'm getting the audit but the relational user is null
see the following image
![Screen Shot 2024-06-05 at 12 34 34](https://github.com/owen-it/laravel-auditing/assets/40463737/d1f24b99-f9fc-4f63-85a9-8e1383b61d32)

**Reply by willpower232** [reply]
Your screenshot shows user_id 1, do you have that row in your users table?

**Reply by newave-ws** [reply]
Of course, it is my main user.

**Reply by willpower232** [reply, marked as answer]
Oh hang on, because you've requested specific columns, you've missed off user_type.

Some simpler version of your code would be

```php
$query = $modelInstance->query()
    ->where('id', $request->id)
    ->with([
        'audits' => function ($query) { $query->orderBy('id', 'desc'); },
        'audits.user',
    ]);
```

**Reply by newave-ws** [reply]
Amazing, thank you for your insight.
I'm getting the user with the audits, can you explain what i have done wrong?

**Reply by newave-ws** [reply]
And can you answer another problem that i don't managed to solve?

**Reply by willpower232** [reply]
I eventually remembered that the user model was a morph relationship

https://github.com/owen-it/laravel-auditing/blob/22682dd5c800aab90ef261530098625e26a66709/src/Audit.php#L48-L53

As your initial code sample included `select()`, you were accidentally not selecting both fields in the database so when Laravel received the data it was not able to look up the user models.

Is it a related problem or should it be a different discussion?

**Reply by newave-ws** [reply]
First, can i get the user specific fields like name and email instead of all the data? 
second, I've already opened discussion about that but i didn't got the answer yet? [https://github.com/owen-it/laravel-auditing/discussions/927](https://github.com/owen-it/laravel-auditing/discussions/927)

**Reply by willpower232** [reply]
I presume you can do

```php
'audits.user' => function ($query) { $query->select('id', 'name', 'email'); },
```

alternatively, you could prevent the data from being output with a [resource](https://laravel.com/docs/11.x/eloquent-resources)

**Reply by newave-ws** [reply]
I'm always forget to add the id and then i get null, i'm embarrassed
